src/Telemetry.h

```diff
@@ -0,0 +1,100 @@
+#ifndef HALIDE_TELEMETRY_H_
+#define HALIDE_TELEMETRY_H_
+
+/** \file
+ * Defines an interface used to gather compile-time information, stats, etc
+ * for use in evaluating internal Halide compilation rules and efficiency.
+ *
+ * There is a 'standard' implementation that simply logs information to
+ * a stream, but the entire implementation can be replaced to log in arbitrary ways.
+ */
```

Comment:
"Telemetry" has some spooky connotations, so we should make it super clear here that this does nothing unless you ask for the telemetry output in a generator, in which case it writes it to a file. In no circumstances does it ever send any information about your code to anyone else, e.g. over the internet.

Reply:
SGTM

Reply:
"compiler tracing" or "bug report helper" or something

Reply:
@alexreinking not sure where you are suggesting those phrases be inserted

Reply:
Oh, sorry, I was just suggesting some was to rephrase "telemetry" without the aforementioned spooky connotations. It would be worth explaining the feature in the Generator help text, now that you mention it.

Reply:
I wonder if this would be better named simply something to do with logging, (and we could consider routing more of our logging/error reporting through it later) rather than the spooky sounding telemetry.

Similarly, "anonymize" might be better expressed as "obfuscate".

Reply:
I'm fine with whatever nomenclature we settle on -- this was the first name that sprang to mind. (spookiness didn't occur to me but since multiple other people say it does, it does :-)

Something along the lines of "CompilerLogger" maybe? (Not very terse but maybe less spooky sounding I guess.) Other specific suggestions before this thing lands?

Re: 'Obfuscate', SGTM.
-

Reply:
Update: going with `CompilerLogger` for now.